pkg/api/rest/request.go

```diff
@@ -0,0 +1,401 @@
+package rest
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	terrors "github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/ticdc/cdc/model"
+	cerrors "github.com/pingcap/ticdc/pkg/errors"
+	"github.com/pingcap/ticdc/pkg/httputil"
+	"github.com/pingcap/ticdc/pkg/retry"
+	"go.uber.org/zap"
+	"golang.org/x/time/rate"
+)
+
+const (
+	defaultBackoffBaseDelayInMs = 500
+	defaultBackoffMaxDelayInMs  = 30 * 1000
+	// The write operations other than 'GET' are not idempotent,
+	// so we only try one time in request.Do() and let users specify the retrying behaviour
+	defaultMaxRetries = 1
+)
+
+// Request allows for building up a request to cdc server in a chained fasion.
+// Any errors are stored until the end of your call, so you only have to
+// check once.
+type Request struct {
+	c *RESTClient
+	// TODO: implement client-side throttling
+	rateLimiter *rate.Limiter
+	timeout     time.Duration
+
+	// generic components accessible via method setters
+	verb       string
+	pathPrefix string
+	params     url.Values
+	headers    http.Header
+
+	// retry options
+	backoffBaseDelay time.Duration
+	backoffMaxDelay  time.Duration
+	maxRetries       int64
+
+	//output
+	err  error
+	body io.Reader
+}
+
+// NewRequest creates a new request.
+func NewRequest(c *RESTClient) *Request {
+	var pathPrefix string
+	if c.base != nil {
+		pathPrefix = path.Join("/", c.base.Path, c.versionedAPIPath)
+	} else {
+		pathPrefix = path.Join("/", c.versionedAPIPath)
+	}
+
+	var timeout time.Duration
+	if c.Client != nil {
+		timeout = c.Client.Timeout
+	}
+
+	r := &Request{
+		c:          c,
+		timeout:    timeout,
+		pathPrefix: pathPrefix,
+		maxRetries: 1,
+	}
+	r.SetHeader("Accept", "application/json")
+	return r
+}
+
+// NewRequestWithClient creates a Request with an embedded RESTClient for test.
+func NewRequestWithClient(base *url.URL, versionedAPIPath string, client *httputil.Client) *Request {
+	return NewRequest(&RESTClient{
+		base:             base,
+		versionedAPIPath: versionedAPIPath,
+		Client:           client,
+	})
+}
+
+// Prefix adds segments to the begining of request url.
+func (r *Request) Prefix(segments ...string) *Request {
+	if r.err != nil {
+		return r
+	}
+
+	r.pathPrefix = path.Join(r.pathPrefix, path.Join(segments...))
+	return r
+}
+
+// URI sets the server relative URI.
+func (r *Request) URI(uri string) *Request {
+	if r.err != nil {
+		return r
+	}
+	u, err := url.Parse(uri)
+	if err != nil {
+		r.err = err
```

Comment:
Maybe we can require a `*url.URL` as the parameter, to avoid having to parse it here?

Reply:
The `URI()` function is used in part 2. https://github.com/pingcap/tiflow/blob/27643133813709878ba260166f7a959a65564f63/pkg/api/v1/changefeed.go#L36-L44